File: jarvis/beacon.py
"""UDP auto-discovery beacon so 1bit Mobile finds this jarvis server on the
LAN without the user typing in an IP address -- matches the
`BeaconListenerService` on port 13305 the app already listens on
(`{"service": "1bit", "hostname": ..., "url": ...}` broadcast datagrams).
"""
import json
import socket
import threading
import time

BEACON_PORT = 13305


def _lan_ip():
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        s.connect(("8.8.8.8", 80))
        return s.getsockname()[0]
    except Exception:
        return "127.0.0.1"
    finally:
        s.close()


def start_beacon(port, interval=3.0):
    """Start a daemon thread broadcasting this server's address every `interval`s."""
    url = f"http://{_lan_ip()}:{port}"
    hostname = socket.gethostname()
    payload = json.dumps({"service": "1bit", "hostname": hostname, "url": url}).encode()

    def _run():
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
        while True:
            try:
                s.sendto(payload, ("<broadcast>", BEACON_PORT))
            except Exception:
                pass
            time.sleep(interval)

    t = threading.Thread(target=_run, daemon=True)
    t.start()
    print(f"  Beacon: advertising {url} on UDP :{BEACON_PORT}")
    return t

File: tools/jarvis/beacon.cpp
#include "beacon.h"

#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <unistd.h>

#include <chrono>
#include <cstdio>
#include <nlohmann/json.hpp>
#include <string>
#include <thread>

using json = nlohmann::json;

namespace jarvis {
namespace {

std::string lan_ip() {
    int sock = socket(AF_INET, SOCK_DGRAM, 0);
    if (sock < 0) return "127.0.0.1";

    sockaddr_in addr{};
    addr.sin_family = AF_INET;
    addr.sin_port = htons(80);
    inet_pton(AF_INET, "8.8.8.8", &addr.sin_addr);

    // connect() on a UDP socket doesn't send a packet — it just picks the
    // outbound-facing local address for getsockname(), same trick the
    // original Python used.
    std::string result = "127.0.0.1";
    if (connect(sock, (sockaddr*)&addr, sizeof(addr)) == 0) {
        sockaddr_in local{};
        socklen_t len = sizeof(local);
        if (getsockname(sock, (sockaddr*)&local, &len) == 0) {
            char buf[INET_ADDRSTRLEN];
            if (inet_ntop(AF_INET, &local.sin_addr, buf, sizeof(buf))) result = buf;
        }
    }
    close(sock);
    return result;
}

std::string hostname() {
    char buf[256];
    if (gethostname(buf, sizeof(buf)) == 0) return buf;
    return "unknown";
}

} // namespace

void start_beacon(int server_port, double interval_seconds) {
    std::string url = "http://" + lan_ip() + ":" + std::to_string(server_port);
    std::string host = hostname();
    json payload = {{"service", "1bit"}, {"hostname", host}, {"url", url}};
    std::string payload_str = payload.dump();

    fprintf(stderr, "Beacon: advertising %s on UDP :%d\n", url.c_str(), kBeaconPort);

    std::thread([payload_str, interval_seconds]() {
        int sock = socket(AF_INET, SOCK_DGRAM, 0);
        if (sock < 0) return;
        int broadcast_enable = 1;
        setsockopt(sock, SOL_SOCKET, SO_BROADCAST, &broadcast_enable, sizeof(broadcast_enable));

        sockaddr_in dest{};
        dest.sin_family = AF_INET;
        dest.sin_port = htons(kBeaconPort);
        dest.sin_addr.s_addr = INADDR_BROADCAST; // 255.255.255.255

        while (true) {
            sendto(sock, payload_str.data(), payload_str.size(), 0, (sockaddr*)&dest, sizeof(dest));
            std::this_thread::sleep_for(std::chrono::milliseconds((long)(interval_seconds * 1000)));
        }
    }).detach();
}

} // namespace jarvis
